File: ben/prolog/P54-69-trees/p62.pl
% Collect the internal nodes of a binary tree in a list
% An internal node of a binary tree has either one or two non-empty successors. Write a predicate internals/2 to collect them in a list.

internals(nil, []).
internals(t(_, nil, nil), []) :- !.
internals(t(X, L, R), [X|N]) :-
    internals(L, NL), 
    internals(R, NR),
    append(NL, NR, N).

File: ben/haskell/Q61-69-binary-trees-cont/p62.hs
{-
Collect the internal nodes of a binary tree in a list

An internal node of a binary tree has either one or two non-empty successors. Write a predicate internals/2 to collect them in a list.
-}

data Tree a = Empty | Branch a (Tree a) (Tree a)
              deriving (Show, Eq)

tree4 = Branch 1 (Branch 2 Empty (Branch 4 Empty Empty))
                 (Branch 2 Empty Empty)

internals :: Tree a -> [a]
internals Empty = []
internals (Branch _ Empty Empty) = []
internals (Branch x l r) = [x] ++ internals l ++ internals r
